# `censusdata` example

Docs: https://jtleider.github.io/censusdata/

```shell
pip install --user censusdata
```

In [3]:
import censusdata
import pandas as pd

In [7]:
# the last five labels that have the word unemploy in them
censusdata.search(src='acs1', year = 2017, field = 'label', criterion = 'unemploy' )[-5:]

[('C23002I_008E',
  'SEX BY AGE BY EMPLOYMENT STATUS FOR THE POPULATION 16 YEARS AND OVER (HISPANIC OR LATINO)',
  'Estimate!!Total!!Male!!16 to 64 years!!In labor force!!Civilian!!Unemployed'),
 ('C23002I_013E',
  'SEX BY AGE BY EMPLOYMENT STATUS FOR THE POPULATION 16 YEARS AND OVER (HISPANIC OR LATINO)',
  'Estimate!!Total!!Male!!65 years and over!!In labor force!!Unemployed'),
 ('C23002I_021E',
  'SEX BY AGE BY EMPLOYMENT STATUS FOR THE POPULATION 16 YEARS AND OVER (HISPANIC OR LATINO)',
  'Estimate!!Total!!Female!!16 to 64 years!!In labor force!!Civilian!!Unemployed'),
 ('C23002I_026E',
  'SEX BY AGE BY EMPLOYMENT STATUS FOR THE POPULATION 16 YEARS AND OVER (HISPANIC OR LATINO)',
  'Estimate!!Total!!Female!!65 years and over!!In labor force!!Unemployed'),
 ('C23004_006E',
  'WORK STATUS IN THE PAST 12 MONTHS BY EMPLOYMENT STATUS FOR THE CIVILIAN POPULATION 65 YEARS AND OVER',
  'Estimate!!Total!!Did not work in the past 12 months!!Unemployed')]

In [9]:
# you can search by "concept"
censusdata.search(src = 'acs1', year = 2015, field = 'concept', criterion = 'education')[-5:]

[('C17018_009M',
  'C17018.  POVERTY STATUS IN THE PAST 12 MONTHS OF FAMILIES BY EDUCATIONAL ATTAINMENT OF HOUSEHOLDER',
  'Margin of Error for!!Income in the past 12 months at or above poverty level:!!High school graduate (includes equivalency)'),
 ('C17018_010E',
  'C17018.  POVERTY STATUS IN THE PAST 12 MONTHS OF FAMILIES BY EDUCATIONAL ATTAINMENT OF HOUSEHOLDER',
  "Income in the past 12 months at or above poverty level:!!Some college, associate's degree"),
 ('C17018_010M',
  'C17018.  POVERTY STATUS IN THE PAST 12 MONTHS OF FAMILIES BY EDUCATIONAL ATTAINMENT OF HOUSEHOLDER',
  "Margin of Error for!!Income in the past 12 months at or above poverty level:!!Some college, associate's degree"),
 ('C17018_011E',
  'C17018.  POVERTY STATUS IN THE PAST 12 MONTHS OF FAMILIES BY EDUCATIONAL ATTAINMENT OF HOUSEHOLDER',
  "Income in the past 12 months at or above poverty level:!!Bachelor's degree or higher"),
 ('C17018_011M',
  'C17018.  POVERTY STATUS IN THE PAST 12 MONTHS OF FAMILIES BY EDU

In [10]:
# not all years have same fields
censusdata.search(src = 'acs1', year = 2017, field = 'concept', criterion = 'education')[-5:]

KeyError: 'concept'

In [11]:
# you can grab tables if you know what you are looking for
censusdata.censustable(src = 'acs5', year = 2015, table = 'B15003')

OrderedDict([('B15003_001E',
              {'label': 'Total:',
               'concept': 'B15003.  Educational Attainment for the Population 25 Years and Over',
               'predicateType': 'int'}),
             ('B15003_001M',
              {'label': 'Margin of Error for!!Total:',
               'concept': 'B15003.  Educational Attainment for the Population 25 Years and Over',
               'predicateType': 'int'}),
             ('B15003_002E',
              {'label': 'No schooling completed',
               'concept': 'B15003.  Educational Attainment for the Population 25 Years and Over',
               'predicateType': 'int'}),
             ('B15003_002M',
              {'label': 'Margin of Error for!!No schooling completed',
               'concept': 'B15003.  Educational Attainment for the Population 25 Years and Over',
               'predicateType': 'int'}),
             ('B15003_003E',
              {'label': 'Nursery school',
               'concept': 'B15003.  Educational

In [12]:
#and print them legibly
censusdata.printtable(censusdata.censustable(src = 'acs5', year = 2015, table = 'B15003'))

Variable     | Table                          | Label                                                    | Type 
-------------------------------------------------------------------------------------------------------------------
B15003_001E  | B15003.  Educational Attainmen | Total:                                                   | int  
B15003_002E  | B15003.  Educational Attainmen | No schooling completed                                   | int  
B15003_003E  | B15003.  Educational Attainmen | Nursery school                                           | int  
B15003_004E  | B15003.  Educational Attainmen | Kindergarten                                             | int  
B15003_005E  | B15003.  Educational Attainmen | 1st grade                                                | int  
B15003_006E  | B15003.  Educational Attainmen | 2nd grade                                                | int  
B15003_007E  | B15003.  Educational Attainmen | 3rd grade                                    

It comes with a `.geographies()` module

In [17]:
# geography object - * is wildcard
geo = censusdata.censusgeo(geo = [('state', '*')])
type(geo)

censusdata.censusgeo.censusgeo

In [18]:
# grab all the states in acs1 of year 2017
censusdata.geographies(within = geo, src='acs1', year=2017)

{'Alabama': censusgeo((('state', '01'),)),
 'Alaska': censusgeo((('state', '02'),)),
 'Arizona': censusgeo((('state', '04'),)),
 'Arkansas': censusgeo((('state', '05'),)),
 'California': censusgeo((('state', '06'),)),
 'Colorado': censusgeo((('state', '08'),)),
 'Connecticut': censusgeo((('state', '09'),)),
 'Delaware': censusgeo((('state', '10'),)),
 'District of Columbia': censusgeo((('state', '11'),)),
 'Florida': censusgeo((('state', '12'),)),
 'Georgia': censusgeo((('state', '13'),)),
 'Hawaii': censusgeo((('state', '15'),)),
 'Idaho': censusgeo((('state', '16'),)),
 'Illinois': censusgeo((('state', '17'),)),
 'Indiana': censusgeo((('state', '18'),)),
 'Iowa': censusgeo((('state', '19'),)),
 'Kansas': censusgeo((('state', '20'),)),
 'Kentucky': censusgeo((('state', '21'),)),
 'Louisiana': censusgeo((('state', '22'),)),
 'Maine': censusgeo((('state', '23'),)),
 'Maryland': censusgeo((('state', '24'),)),
 'Massachusetts': censusgeo((('state', '25'),)),
 'Michigan': censusgeo((('stat

In [23]:
# grab all counties in state X from acs1 in 2017
geo = censusdata.censusgeo([("state", "06"), ("county", "*")])
censusdata.geographies(within=geo, src = 'acs1', year = 2017)

{'Alameda County, California': censusgeo((('state', '06'), ('county', '001'))),
 'Butte County, California': censusgeo((('state', '06'), ('county', '007'))),
 'Contra Costa County, California': censusgeo((('state', '06'), ('county', '013'))),
 'El Dorado County, California': censusgeo((('state', '06'), ('county', '017'))),
 'Fresno County, California': censusgeo((('state', '06'), ('county', '019'))),
 'Humboldt County, California': censusgeo((('state', '06'), ('county', '023'))),
 'Imperial County, California': censusgeo((('state', '06'), ('county', '025'))),
 'Kern County, California': censusgeo((('state', '06'), ('county', '029'))),
 'Kings County, California': censusgeo((('state', '06'), ('county', '031'))),
 'Lake County, California': censusgeo((('state', '06'), ('county', '033'))),
 'Los Angeles County, California': censusgeo((('state', '06'), ('county', '037'))),
 'Madera County, California': censusgeo((('state', '06'), ('county', '039'))),
 'Marin County, California': censusgeo(

# Exercise

education variables

In [27]:
censusdata.printtable(censusdata.censustable(src = 'acs5', year = 2017, table = 'B15003'))

Variable     | Table                          | Label                                                    | Type 
-------------------------------------------------------------------------------------------------------------------
B15003_001E  | EDUCATIONAL ATTAINMENT FOR THE | !! Estimate Total                                        | int  
B15003_002E  | EDUCATIONAL ATTAINMENT FOR THE | !! !! Estimate Total No schooling completed              | int  
B15003_003E  | EDUCATIONAL ATTAINMENT FOR THE | !! !! Estimate Total Nursery school                      | int  
B15003_004E  | EDUCATIONAL ATTAINMENT FOR THE | !! !! Estimate Total Kindergarten                        | int  
B15003_005E  | EDUCATIONAL ATTAINMENT FOR THE | !! !! Estimate Total 1st grade                           | int  
B15003_006E  | EDUCATIONAL ATTAINMENT FOR THE | !! !! Estimate Total 2nd grade                           | int  
B15003_007E  | EDUCATIONAL ATTAINMENT FOR THE | !! !! Estimate Total 3rd grade               

Let's get % of population with at least a bachelor's degree

- Bachelor's degree: `B15003_022E`
- Master's degree: `B15003_023E`
- Professional school degree: `B15003_024E`
- Doctorate degree: `B15003_025E`
- total: `B15003_001E`

$$Share\ of\ Pop.\ with\ at\ least\ a\ bachelor's\ degree = \frac{B15003\_022E + B15003\_023E + B15003\_024E + B15003\_025E}{B15003\_001E}$$

In [31]:
my_vars = ["B15003_022E", "B15003_023E", "B15003_024E", "B15003_025E", "B15003_001E"]

Grab your geographies

In [30]:
geography_of_interest = [("state", "06"), ("county","*")]
my_geo = censusdata.censusgeo(geo = geography_of_interest)

# Download data

In [29]:
censusdata.download?

Signature:
censusdata.download(
    src,
    year,
    geo,
    var,
    key=None,
    tabletype='detail',
    endpt='',
)
Docstring:
Download data from Census API.

Args:
        src (str): Census data source: 'acs1' for ACS 1-year estimates, 'acs5' for ACS 5-year estimates, 'acs3' for
                ACS 3-year estimates, 'acsse' for ACS 1-year supplemental estimates, 'sf1' for SF1 data.
        year (int): Year of data.
        geo (censusgeo): Geographies for which to download data.
        var (list of str): Census variables to download.
        key (str, optional): Census API key.
        tabletype (str, optional): Type of table from which variables are drawn (only applicable to ACS data). Options are 'detail' (detail tables),
                'subject' (subject tables), 'profile' (data profile tables), 'cprofile' (comparison profile tables).
        endpt (str, optional): Allows override of whether old or new API endpoint is used. Specify
                'old' for old, 'new' for 

In [32]:
data = censusdata.download(src = 'acs5', year = 2017, geo = my_geo, var = my_vars)

In [33]:
data.head()

B15003_022E  B15003_023E  \
Modoc County, California: Summary level: 050, s...          859          227   
Merced County, California: Summary level: 050, ...        14006         5063   
Lake County, California: Summary level: 050, st...         4588         1678   
Mariposa County, California: Summary level: 050...         2017          933   
Yuba County, California: Summary level: 050, st...         5573         1565   

                                                    B15003_024E  B15003_025E  \
Modoc County, California: Summary level: 050, s...           50            0   
Merced County, California: Summary level: 050, ...         1337         1135   
Lake County, California: Summary level: 050, st...          577          273   
Mariposa County, California: Summary level: 050...          253          132   
Yuba County, California: Summary level: 050, st...          406           84   

                                                    B15003_001E  
Modoc County, California: Summary level: 050, s...         6852  
Merced County, California: Summary level: 050, ...       156562  
Lake County, California: Summary level: 050, st...        46386  
Mariposa County, California: Summary level: 050...        13656  
Yuba County, California: Summary level: 050, st...        46582

In [34]:
rename_vars = {
    "B15003_022E": "Bachelors", 
    "B15003_023E": "Masters", 
    "B15003_024E": "Professional degree", 
    "B15003_025E": "Doctorates",
    "B15003_001E": "Total",
}

In [36]:
# need to do this because rename returns dataframe
data = data.rename(mapper = rename_vars, axis = 1)

In [37]:
data.head()

Bachelors  Masters  \
Modoc County, California: Summary level: 050, s...        859      227   
Merced County, California: Summary level: 050, ...      14006     5063   
Lake County, California: Summary level: 050, st...       4588     1678   
Mariposa County, California: Summary level: 050...       2017      933   
Yuba County, California: Summary level: 050, st...       5573     1565   

                                                    Professional degree  \
Modoc County, California: Summary level: 050, s...                   50   
Merced County, California: Summary level: 050, ...                 1337   
Lake County, California: Summary level: 050, st...                  577   
Mariposa County, California: Summary level: 050...                  253   
Yuba County, California: Summary level: 050, st...                  406   

                                                    Doctorates   Total  
Modoc County, California: Summary level: 050, s...           0    6852  
Merced County, California: Summary level: 050, ...        1135  156562  
Lake County, California: Summary level: 050, st...         273   46386  
Mariposa County, California: Summary level: 050...         132   13656  
Yuba County, California: Summary level: 050, st...          84   46582

In [38]:
data['share of bachelors+'] = (data['Bachelors'] + data['Masters'] + data['Professional degree'] + data['Doctorates']) / data['Total']

In [40]:
data.head()

Bachelors  Masters  \
Modoc County, California: Summary level: 050, s...        859      227   
Merced County, California: Summary level: 050, ...      14006     5063   
Lake County, California: Summary level: 050, st...       4588     1678   
Mariposa County, California: Summary level: 050...       2017      933   
Yuba County, California: Summary level: 050, st...       5573     1565   

                                                    Professional degree  \
Modoc County, California: Summary level: 050, s...                   50   
Merced County, California: Summary level: 050, ...                 1337   
Lake County, California: Summary level: 050, st...                  577   
Mariposa County, California: Summary level: 050...                  253   
Yuba County, California: Summary level: 050, st...                  406   

                                                    Doctorates   Total  \
Modoc County, California: Summary level: 050, s...           0    6852   
Merced County, California: Summary level: 050, ...        1135  156562   
Lake County, California: Summary level: 050, st...         273   46386   
Mariposa County, California: Summary level: 050...         132   13656   
Yuba County, California: Summary level: 050, st...          84   46582   

                                                    share of bachelors+  
Modoc County, California: Summary level: 050, s...             0.165791  
Merced County, California: Summary level: 050, ...             0.137588  
Lake County, California: Summary level: 050, st...             0.153408  
Mariposa County, California: Summary level: 050...             0.244215  
Yuba County, California: Summary level: 050, st...             0.163754

# Advanced data prepping

This data kind of works but ideally we would have something like this

| County |    State   | Summary level | State Code | County Code | Bachelors | Masters | Professional degree | Doctorates | Total | share of bachelors+ |
|:------:|:----------:|:-------------:|:----------:|:-----------:|:---------:|:-------:|:-------------------:|:----------:|:-----:|:-------------------:|
| Modoc  | California |       050     |     06     |     049     |     859   |   227   |          50         |      0     |  6852 |       0.165791      |
| Merced | California |       050     |     06     |     047     |   14006   |  5063   |        1337         |   1135     |156562 |       0.137588      |
| Lake   | California |       050     |     06     |     033     |    4588   |  1678   |         577         |    273     | 56386 |       0.153408      |
|Mariposa| California |       050     |     06     |     043     |    2017   |   933   |         253         |    132     | 13656 |       0.244215      |
|  Yuba  | California |       050     |     06     |     115     |    5573   |  1565   |         406         |     84     | 56582 |       0.163754      |

First step, `.reset_index()`

In [42]:
# this returns dataframe so we need to assign it back to itself
data = data.reset_index()

In [43]:
data.head()

index  Bachelors  Masters  \
0  Modoc County, California: Summary level: 050, ...        859      227   
1  Merced County, California: Summary level: 050,...      14006     5063   
2  Lake County, California: Summary level: 050, s...       4588     1678   
3  Mariposa County, California: Summary level: 05...       2017      933   
4  Yuba County, California: Summary level: 050, s...       5573     1565   

   Professional degree  Doctorates   Total  share of bachelors+  
0                   50           0    6852             0.165791  
1                 1337        1135  156562             0.137588  
2                  577         273   46386             0.153408  
3                  253         132   13656             0.244215  
4                  406          84   46582             0.163754

`.str` accessor - lets you apply string methods to pandas series - however our index is a `censusgeo` object

In [57]:
print(data['index'][0])
print(data['index'][0].hierarchy())

Modoc County, California: Summary level: 050, state:06> county:049
state> county


Save as string

In [58]:
data['index_as_string'] = data['index'].astype(str)

In [59]:
data.head()

index  Bachelors  Masters  \
0  Modoc County, California: Summary level: 050, ...        859      227   
1  Merced County, California: Summary level: 050,...      14006     5063   
2  Lake County, California: Summary level: 050, s...       4588     1678   
3  Mariposa County, California: Summary level: 05...       2017      933   
4  Yuba County, California: Summary level: 050, s...       5573     1565   

   Professional degree  Doctorates   Total  share of bachelors+  \
0                   50           0    6852             0.165791   
1                 1337        1135  156562             0.137588   
2                  577         273   46386             0.153408   
3                  253         132   13656             0.244215   
4                  406          84   46582             0.163754   

                                     index_as_string  
0  Modoc County, California: Summary level: 050, ...  
1  Merced County, California: Summary level: 050,...  
2  Lake County, California: Summary level: 050, s...  
3  Mariposa County, California: Summary level: 05...  
4  Yuba County, California: Summary level: 050, s...

In [68]:
# .str accessor
data['index_as_string'].str.split(":").str[0]

0               Modoc County, California
1              Merced County, California
2                Lake County, California
3            Mariposa County, California
4                Yuba County, California
5        Contra Costa County, California
6                Inyo County, California
7              Lassen County, California
8          Stanislaus County, California
9       Santa Barbara County, California
10             Sonoma County, California
11             Tehama County, California
12           Imperial County, California
13               Mono County, California
14         San Benito County, California
15            Alameda County, California
16         Sacramento County, California
17          El Dorado County, California
18               Napa County, California
19          San Diego County, California
20           Monterey County, California
21      San Francisco County, California
22             Sierra County, California
23             Tulare County, California
24              

### now we know somethings already 
1. these are all in California so `State` and `State Code` are 'California' and '06' respectively for all of them.
2. Summary Level is the same for all: 050

So what we really need is `County` and `County Code`

In [73]:
# rough way 
data['index_as_string'].str.split("County").str[0]

0               Modoc 
1              Merced 
2                Lake 
3            Mariposa 
4                Yuba 
5        Contra Costa 
6                Inyo 
7              Lassen 
8          Stanislaus 
9       Santa Barbara 
10             Sonoma 
11             Tehama 
12           Imperial 
13               Mono 
14         San Benito 
15            Alameda 
16         Sacramento 
17          El Dorado 
18               Napa 
19          San Diego 
20           Monterey 
21      San Francisco 
22             Sierra 
23             Tulare 
24               Yolo 
25           Humboldt 
26             Nevada 
27             Alpine 
28             Madera 
29          Mendocino 
30         Santa Cruz 
31            Trinity 
32        Los Angeles 
33          Riverside 
34     San Bernardino 
35        Santa Clara 
36              Marin 
37             Placer 
38           Siskiyou 
39             Shasta 
40             Solano 
41          Del Norte 
42             Colusa 
43         

In [75]:
data['index_as_string'].str.split("county").str[-1].str.replace(":", "")

0     049
1     047
2     033
3     043
4     115
5     013
6     027
7     035
8     099
9     083
10    097
11    103
12    025
13    051
14    069
15    001
16    067
17    017
18    055
19    073
20    053
21    075
22    091
23    107
24    113
25    023
26    057
27    003
28    039
29    045
30    087
31    105
32    037
33    065
34    071
35    085
36    041
37    061
38    093
39    089
40    095
41    015
42    011
43    021
44    007
45    031
46    063
47    079
48    019
49    029
50    059
51    111
52    009
53    101
54    081
55    109
56    077
57    005
Name: index_as_string, dtype: object

Pandas split has `expand` option

examine first observation

In [80]:
data['index_as_string'][0]

'Modoc County, California: Summary level: 050, state:06> county:049'

The county name is right before "County,"

In [86]:
# gotta chain .str accessor every time you need to apply a string method to a series
data['index_as_string'].str.split("County,").str[0].str.strip()

0               Modoc
1              Merced
2                Lake
3            Mariposa
4                Yuba
5        Contra Costa
6                Inyo
7              Lassen
8          Stanislaus
9       Santa Barbara
10             Sonoma
11             Tehama
12           Imperial
13               Mono
14         San Benito
15            Alameda
16         Sacramento
17          El Dorado
18               Napa
19          San Diego
20           Monterey
21      San Francisco
22             Sierra
23             Tulare
24               Yolo
25           Humboldt
26             Nevada
27             Alpine
28             Madera
29          Mendocino
30         Santa Cruz
31            Trinity
32        Los Angeles
33          Riverside
34     San Bernardino
35        Santa Clara
36              Marin
37             Placer
38           Siskiyou
39             Shasta
40             Solano
41          Del Norte
42             Colusa
43              Glenn
44              Butte
45        

In [87]:
# save it as a colum
data['County'] = data['index_as_string'].str.split("County,").str[0].str.strip()

In [88]:
data.head()

index  Bachelors  Masters  \
0  Modoc County, California: Summary level: 050, ...        859      227   
1  Merced County, California: Summary level: 050,...      14006     5063   
2  Lake County, California: Summary level: 050, s...       4588     1678   
3  Mariposa County, California: Summary level: 05...       2017      933   
4  Yuba County, California: Summary level: 050, s...       5573     1565   

   Professional degree  Doctorates   Total  share of bachelors+  \
0                   50           0    6852             0.165791   
1                 1337        1135  156562             0.137588   
2                  577         273   46386             0.153408   
3                  253         132   13656             0.244215   
4                  406          84   46582             0.163754   

                                     index_as_string    County  
0  Modoc County, California: Summary level: 050, ...     Modoc  
1  Merced County, California: Summary level: 050,...    Merced  
2  Lake County, California: Summary level: 050, s...      Lake  
3  Mariposa County, California: Summary level: 05...  Mariposa  
4  Yuba County, California: Summary level: 050, s...      Yuba

In [91]:
# same with county code
data['index_as_string'].str.split("county:").str[-1].str.strip()

0     049
1     047
2     033
3     043
4     115
5     013
6     027
7     035
8     099
9     083
10    097
11    103
12    025
13    051
14    069
15    001
16    067
17    017
18    055
19    073
20    053
21    075
22    091
23    107
24    113
25    023
26    057
27    003
28    039
29    045
30    087
31    105
32    037
33    065
34    071
35    085
36    041
37    061
38    093
39    089
40    095
41    015
42    011
43    021
44    007
45    031
46    063
47    079
48    019
49    029
50    059
51    111
52    009
53    101
54    081
55    109
56    077
57    005
Name: index_as_string, dtype: object

In [92]:
data['County Code'] = data['index_as_string'].str.split("county:").str[-1].str.strip()

In [93]:
data.head()

index  Bachelors  Masters  \
0  Modoc County, California: Summary level: 050, ...        859      227   
1  Merced County, California: Summary level: 050,...      14006     5063   
2  Lake County, California: Summary level: 050, s...       4588     1678   
3  Mariposa County, California: Summary level: 05...       2017      933   
4  Yuba County, California: Summary level: 050, s...       5573     1565   

   Professional degree  Doctorates   Total  share of bachelors+  \
0                   50           0    6852             0.165791   
1                 1337        1135  156562             0.137588   
2                  577         273   46386             0.153408   
3                  253         132   13656             0.244215   
4                  406          84   46582             0.163754   

                                     index_as_string    County County Code  
0  Modoc County, California: Summary level: 050, ...     Modoc         049  
1  Merced County, California: Summary level: 050,...    Merced         047  
2  Lake County, California: Summary level: 050, s...      Lake         033  
3  Mariposa County, California: Summary level: 05...  Mariposa         043  
4  Yuba County, California: Summary level: 050, s...      Yuba         115

In [95]:
# the rest of the columns are easy to create
data['State'] = 'California'
data['State Code'] = '06'
data['Summary Level'] = '050'

Pandas creates a series with all values = to the string you pass

In [96]:
data.head()

index  Bachelors  Masters  \
0  Modoc County, California: Summary level: 050, ...        859      227   
1  Merced County, California: Summary level: 050,...      14006     5063   
2  Lake County, California: Summary level: 050, s...       4588     1678   
3  Mariposa County, California: Summary level: 05...       2017      933   
4  Yuba County, California: Summary level: 050, s...       5573     1565   

   Professional degree  Doctorates   Total  share of bachelors+  \
0                   50           0    6852             0.165791   
1                 1337        1135  156562             0.137588   
2                  577         273   46386             0.153408   
3                  253         132   13656             0.244215   
4                  406          84   46582             0.163754   

                                     index_as_string    County County Code  \
0  Modoc County, California: Summary level: 050, ...     Modoc         049   
1  Merced County, California: Summary level: 050,...    Merced         047   
2  Lake County, California: Summary level: 050, s...      Lake         033   
3  Mariposa County, California: Summary level: 05...  Mariposa         043   
4  Yuba County, California: Summary level: 050, s...      Yuba         115   

        State State Code Summary Level  
0  California         06           050  
1  California         06           050  
2  California         06           050  
3  California         06           050  
4  California         06           050

In [106]:
# select a group of columns
columns_im_interested_in = ['State', 'State Code', 'County', 'County Code', 'Summary Level', 'share of bachelors+']
data[columns_im_interested_in].head()

State State Code    County County Code Summary Level  \
0  California         06     Modoc         049           050   
1  California         06    Merced         047           050   
2  California         06      Lake         033           050   
3  California         06  Mariposa         043           050   
4  California         06      Yuba         115           050   

   share of bachelors+  
0             0.165791  
1             0.137588  
2             0.153408  
3             0.244215  
4             0.163754

save that to a clean dataframe

In [107]:
clean_data = data[columns_im_interested_in]

# .copy() vs subset

Right now, clean_data is "pointing" (is a `view`) to a subset of data so any changes you do to clean_data aren't _guaranteed_ to work. sometimes they do sometimes they don't. So you should create a .copy() of the data if you plan on modifying it after.

In [108]:
clean_data['State'] = 'CA'

C:\Users\sanchez\AppData\Local\Continuum\anaconda3\envs\pycon2019\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [109]:
clean_data = data[columns_im_interested_in].copy()

In [110]:
clean_data['State'] = 'CA'

No warning

In [111]:
clean_data.head()

State State Code    County County Code Summary Level  share of bachelors+
0    CA         06     Modoc         049           050             0.165791
1    CA         06    Merced         047           050             0.137588
2    CA         06      Lake         033           050             0.153408
3    CA         06  Mariposa         043           050             0.244215
4    CA         06      Yuba         115           050             0.163754

Save data

In [112]:
clean_data.to_csv("../data/interim/censusdata_example.csv", index = False,)